File: Shopify/pair_programming/rate_limiter.cpp
/*
 * Shopify Pair Programming - Rate Limiter
 *
 * Problem (Layer 1):
 *   Implement a rate limiter that allows at most N requests per time window.
 *   isAllowed(userId) -> bool
 *
 * Extensions reported in interviews:
 *   Layer 2: Sliding window (not fixed window).
 *   Layer 3: Per-user AND global rate limits. Redis-backed implementation.
 *   Layer 4 (staff): Distributed rate limiting across multiple API gateway nodes.
 *              Shopify serves Black Friday: 100k req/s. How does this scale?
 *
 * Two main algorithms:
 *   Fixed Window Counter  — simple, cheap, but allows 2x burst at window boundary
 *   Sliding Window Log    — accurate but memory O(requests)
 *   Token Bucket          — smooth, allows controlled bursts, Shopify API uses this
 */

#include <iostream>
#include <unordered_map>
#include <deque>
#include <chrono>
#include <mutex>
#include <string>

using Clock = std::chrono::steady_clock;
using Ms = std::chrono::milliseconds;

// ============================================================
// Layer 1: Fixed Window Counter
// ============================================================

class FixedWindowRateLimiter {
public:
    // maxRequests per windowMs milliseconds
    FixedWindowRateLimiter(int maxRequests, int windowMs)
        : maxRequests(maxRequests), windowMs(windowMs) {}

    bool isAllowed(const std::string& userId) {
        // FIX: the old version computed `time_since_epoch().count() % windowMs`,
        // mixing steady_clock's native ticks (nanoseconds) with a millisecond
        // window -> meaningless boundaries. Convert to ms first, then bucket.
        long long nowMs = std::chrono::duration_cast<Ms>(
            Clock::now().time_since_epoch()).count();
        long long windowId = nowMs / windowMs;

        auto& state = counters[userId];
        if (state.windowId != windowId) {
            state.windowId = windowId;
            state.count = 0;
        }
        if (state.count >= maxRequests) return false;
        state.count++;
        return true;
    }

private:
    int maxRequests, windowMs;
    struct WindowState {
        long long windowId = -1;
        int count = 0;
    };
    std::unordered_map<std::string, WindowState> counters;
};

// ============================================================
// Layer 2: Sliding Window Log (accurate)
// ============================================================

class SlidingWindowRateLimiter {
public:
    SlidingWindowRateLimiter(int maxRequests, int windowMs)
        : maxRequests(maxRequests), windowMs(windowMs) {}

    bool isAllowed(const std::string& userId) {
        auto now = Clock::now();
        auto& log = requestLog[userId];

        // Evict entries outside the window
        auto cutoff = now - Ms(windowMs);
        while (!log.empty() && log.front() <= cutoff) {
            log.pop_front();
        }

        if ((int)log.size() >= maxRequests) return false;
        log.push_back(now);
        return true;
    }

private:
    int maxRequests, windowMs;
    std::unordered_map<std::string, std::deque<Clock::time_point>> requestLog;
};

// ============================================================
// Layer 3: Token Bucket (allows controlled bursts — Shopify API model)
// ============================================================

class TokenBucketRateLimiter {
public:
    // refillRate = tokens per second, bucketSize = max burst
    TokenBucketRateLimiter(double refillRatePerSec, int bucketSize)
        : refillRatePerSec(refillRatePerSec), bucketSize(bucketSize) {}

    bool isAllowed(const std::string& userId, int tokens = 1) {
        std::lock_guard<std::mutex> lock(mu);
        auto now = Clock::now();
        auto& bucket = buckets[userId];

        if (bucket.lastRefill == Clock::time_point{}) {
            bucket.tokens = bucketSize;
            bucket.lastRefill = now;
        }

        double elapsed = std::chrono::duration<double>(now - bucket.lastRefill).count();
        bucket.tokens = std::min((double)bucketSize, bucket.tokens + elapsed * refillRatePerSec);
        bucket.lastRefill = now;

        if (bucket.tokens < tokens) return false;
        bucket.tokens -= tokens;
        return true;
    }

private:
    double refillRatePerSec;
    int bucketSize;
    std::mutex mu;
    struct Bucket {
        double tokens = 0;
        Clock::time_point lastRefill;
    };
    std::unordered_map<std::string, Bucket> buckets;
};

// ============================================================
// Staff Layer 4 Discussion Points
// ============================================================
/*
 * Q: Shopify API Gateway, 100k req/s, multiple nodes. How do you rate limit?
 *
 * Problem: if each node has its own counter, a user can hit N*numNodes requests.
 *
 * Solution A: Centralized Redis counter
 *   - INCR key (atomic), EXPIRE on first set
 *   - Adds ~1ms network RTT per request — acceptable for API calls
 *   - Redis single-threaded for counters = no race conditions
 *   - Risk: Redis becomes SPOF -> use Redis Cluster or Sentinel
 *
 * Solution B: Approximate local rate limiting
 *   - Each node limits to N/numNodes
 *   - No network overhead, but requires knowing numNodes (use service discovery)
 *   - Allows some over-limit when nodes unbalanced
 *
 * Solution C: Gossip protocol (eventual consistency)
 *   - Nodes share counter updates asynchronously
 *   - Very low latency, slightly over-allows during propagation delay
 *   - Good for "soft" rate limits (analytics ingestion)
 *
 * Q: Shopify's actual approach (public knowledge):
 *   - Uses leaky bucket / token bucket model
 *   - API key gets a "bucket" of call credits
 *   - Bucket refills at a rate based on plan tier
 *   - Headers returned: X-Shopify-Shop-Api-Call-Limit: current/max
 *
 * Q: What to monitor?
 *   - Rate limit hit rate per merchant (spikes = possible attack or bad client)
 *   - Redis latency p99
 *   - False positive rate (legitimate users throttled)
 */

// ============================================================
// Tests
// ============================================================

void runTests() {
    {
        SlidingWindowRateLimiter limiter(3, 1000); // 3 req per second
        std::cout << "Sliding window:" << std::endl;
        for (int i = 0; i < 4; i++) {
            bool ok = limiter.isAllowed("user1");
            std::cout << "  Request " << (i+1) << ": " << (ok ? "ALLOWED" : "DENIED") << std::endl;
        }
        // First 3 allowed, 4th denied
    }

    {
        TokenBucketRateLimiter limiter(2.0, 5); // 2 tokens/sec, burst up to 5
        std::cout << "Token bucket (burst test):" << std::endl;
        for (int i = 0; i < 6; i++) {
            bool ok = limiter.isAllowed("user1");
            std::cout << "  Request " << (i+1) << ": " << (ok ? "ALLOWED" : "DENIED") << std::endl;
        }
        // First 5 allowed (full bucket), 6th denied
    }
}

int main() {
    runTests();
    return 0;
}

File: Shopify/java/RateLimiter.java
import java.util.*;
import java.util.function.LongSupplier;

/**
 * Shopify Pair Programming - Rate Limiter
 * Reported across Blind, 2024-2025.
 *
 * Java 17.  Run:  javac RateLimiter.java && java RateLimiter
 *
 * Three algorithms, increasing sophistication - all behind one Limiter interface:
 *   FixedWindow   - cheap, but allows up to 2x burst at a window edge
 *   SlidingWindow - accurate, O(requests-in-window) memory
 *   TokenBucket   - smooth + controlled bursts (Shopify's public API model)
 *
 * Time is INJECTED (LongSupplier of millis) so tests are deterministic - no sleeps.
 */
public class RateLimiter {

    interface Limiter { boolean allow(String key); }

    /** Fixed window: bucket by integer window-id. (No unit mixing - the classic bug.) */
    static class FixedWindow implements Limiter {
        private final int max;
        private final long windowMs;
        private final LongSupplier clock;
        private final Map<String, long[]> state = new HashMap<>(); // key -> {windowId, count}

        FixedWindow(int max, long windowMs, LongSupplier clock) {
            this.max = max; this.windowMs = windowMs; this.clock = clock;
        }
        public boolean allow(String key) {
            long windowId = clock.getAsLong() / windowMs;
            long[] s = state.computeIfAbsent(key, k -> new long[]{windowId, 0});
            if (s[0] != windowId) { s[0] = windowId; s[1] = 0; }
            if (s[1] >= max) return false;
            s[1]++;
            return true;
        }
    }

    /** Sliding window log: exact count over the trailing window. */
    static class SlidingWindow implements Limiter {
        private final int max;
        private final long windowMs;
        private final LongSupplier clock;
        private final Map<String, Deque<Long>> log = new HashMap<>();

        SlidingWindow(int max, long windowMs, LongSupplier clock) {
            this.max = max; this.windowMs = windowMs; this.clock = clock;
        }
        public boolean allow(String key) {
            long now = clock.getAsLong();
            Deque<Long> dq = log.computeIfAbsent(key, k -> new ArrayDeque<>());
            long cutoff = now - windowMs;
            while (!dq.isEmpty() && dq.peekFirst() <= cutoff) dq.pollFirst();
            if (dq.size() >= max) return false;
            dq.addLast(now);
            return true;
        }
    }

    /** Token bucket: refill at a steady rate, spend one token per request. */
    static class TokenBucket implements Limiter {
        private final double ratePerMs;
        private final double capacity;
        private final LongSupplier clock;
        private final Map<String, double[]> buckets = new HashMap<>(); // key -> {tokens, lastRefillMs}

        TokenBucket(double refillPerSecond, double capacity, LongSupplier clock) {
            this.ratePerMs = refillPerSecond / 1000.0;
            this.capacity = capacity;
            this.clock = clock;
        }
        public boolean allow(String key) {
            long now = clock.getAsLong();
            double[] b = buckets.computeIfAbsent(key, k -> new double[]{capacity, now});
            double elapsed = now - b[1];
            b[0] = Math.min(capacity, b[0] + elapsed * ratePerMs);   // refill
            b[1] = now;
            if (b[0] < 1.0) return false;
            b[0] -= 1.0;
            return true;
        }
    }

    /* ---- Staff discussion ----
     * Distributed (100k req/s across N gateways): per-node counters let a user do
     *   N x limit. Options: (A) central Redis INCR+EXPIRE (atomic, ~1ms RTT);
     *   (B) local limit of max/N (no network, approximate); (C) gossip (eventual).
     * Shopify's API uses a leaky/token bucket per API key and returns
     *   X-Shopify-Shop-Api-Call-Limit: used/max.
     * Monitor: throttle rate per merchant (spikes = attack or bad client), Redis p99.
     */

    public static void main(String[] args) {
        long[] now = {0};
        LongSupplier clock = () -> now[0];

        FixedWindow fw = new FixedWindow(3, 1000, clock);
        now[0] = 0;
        check("fw 1-3 allow",   fw.allow("u") && fw.allow("u") && fw.allow("u"));
        check("fw 4 blocked",   !fw.allow("u"));
        now[0] = 1000;                                  // next window
        check("fw new window",  fw.allow("u"));

        SlidingWindow sw = new SlidingWindow(3, 1000, clock);
        now[0] = 0;
        sw.allow("u"); sw.allow("u"); sw.allow("u");
        check("sw 4 blocked",   !sw.allow("u"));
        now[0] = 1001;                                  // first three slide out
        check("sw after slide", sw.allow("u"));

        TokenBucket tb = new TokenBucket(2.0, 5, clock); // 2/sec, burst 5
        now[0] = 0;
        check("tb burst 5", tb.allow("u") && tb.allow("u") && tb.allow("u")
                          && tb.allow("u") && tb.allow("u"));
        check("tb 6 blocked",   !tb.allow("u"));
        now[0] = 1000;                                  // +1s => +2 tokens
        check("tb refill 2",    tb.allow("u") && tb.allow("u"));
        check("tb empty again", !tb.allow("u"));
    }

    static void check(String name, boolean pass) {
        System.out.printf("%-16s %s%n", name, pass ? "PASS" : "FAIL");
    }
}
